src/backend/joanie/payment/api.py

```diff
@@ -0,0 +1,189 @@
+"""
+API endpoints
+"""
+import logging
+
+from django.db import IntegrityError, transaction
+from django.http import HttpResponse
+
+from rest_framework import mixins, permissions, viewsets
+from rest_framework.decorators import api_view, permission_classes
+from rest_framework.response import Response
+
+from joanie.core.enums import ORDER_STATE_PENDING
+from joanie.core.models import Course, Order, Product, User
+from joanie.payment.models import Invoice
+
+from . import exceptions, get_payment_backend, serializers
+from .models import CreditCard
+
+logger = logging.getLogger(__name__)
+
+
+# pylint: disable=too-many-return-statements
+@api_view(["POST"])
+@permission_classes([permissions.IsAuthenticated])
+@transaction.atomic()
+def create_payment(request):
+    """
+    Extract product_id, course_code and billing_address from request
+    get or create related order then create a payment from payment provider
+    and finally return a response containing payment information.
+    """
+
+    payment_backend = get_payment_backend()
+    product_id = request.data.get("product_id")
+    course_code = request.data.get("course_code")
+    billing_address = request.data.get("billing_address")
+    credit_card_id = request.data.get("credit_card_id")
+    user = User.objects.get_or_create(username=request.user.username)[0]
+
+    if product_id is None:
+        return Response({"product_id": "This field is required."}, status=400)
+
+    if course_code is None:
+        return Response({"course_code": "This field is required."}, status=400)
+
+    if billing_address is None:
+        return Response({"billing_address": "This field is required."}, status=400)
+
+    try:
+        course = Course.objects.get(code=course_code)
+        product = Product.objects.get(uid=product_id)
+    except Course.DoesNotExist:
+        return Response(f"Course {course_code} does not exist.", status=404)
+    except Product.DoesNotExist:
+        return Response(f"Product {product_id} does not exist.", status=404)
+
+    # Create a new order
+    try:
+        order = Order.objects.get(
+            product=product, course=course, owner=user, is_canceled=False
+        )
+
+        if order.state != ORDER_STATE_PENDING:
+            raise IntegrityError(f"{user.username} already owned product {product_id}.")
+        if order.state == ORDER_STATE_PENDING:
+            raise IntegrityError(
+                f"{user.username} already tried to purchase product {product_id}."
+            )
```

Comment:
Hmm if you look at the whole try/except, you'll see that I have to put this code here because the statement below `except Order.DoesNotExist` is in charge to create an order. So if I check after the try/except the state of order I always get a pending state.

Furthermore, this code already returns a 400 not a 500. I raise an IntegrityError that is catched by the try/except. Then I log details error then I returned a bad request with a more generic error message.